### ISSUE TITLE
Feature: add a read-only normalization compatibility report

### ISSUE THREAD

**Issue by TheRealBaoBao** [Physical-Intelligence/openpi#1025]
When fine-tuning pi0 or pi0.5 on a custom robot dataset, users must choose between computing new normalization statistics and reusing statistics from a pretrained checkpoint. Reusing checkpoint statistics can help when the robot and action-space definition match pretraining, but OpenPI currently does not provide a read-only way to measure whether the configured statistics are compatible with the post-transform training data.

A mismatch can remain syntactically valid while producing severe clipping, out-of-distribution normalized values, or unstable training. Related reports include #773, #814, and #832.

Proposed feature

Add a small read-only normalization compatibility report, either as a new script or as an optional mode of scripts/compute_norm_stats.py, depending on the maintainers’ preferred interface.

Example:

uv run scripts/check_norm_stats.py \
  --config-name pi05_droid_finetune \
  --max-frames 5000

The command would use OpenPI’s existing configuration and data pipeline so values are inspected after the configured repack and data transforms. It would then compare sampled state and actions against the normalization statistics selected by the configuration.

Example output:

actions
  data shape:                  (..., 8)
  norm-stat shape:             (8,)
  non-finite values:           0
  near-constant dimensions:    [7]
  outside checkpoint q01-q99:  17.4%
  largest mismatch:            dimension 4
  result:                      WARNING

state
  data shape:                  (..., 8)
  norm-stat shape:             (8,)
  non-finite values:           0
  outside checkpoint q01-q99:  2.1%
  result:                      PASS

The initial checks would be intentionally limited to:

Missing keys and incompatible final dimensions
NaN and Inf values
Near-zero-variance dimensions
Fraction of sampled values outside the configured q01–q99 range
Per-dimension summaries identifying the largest mismatch

Thresholds could initially be informational rather than blocking training.

Non-goals
No mutation, filtering, or rewriting of datasets
No automatic choice between pretrained and newly computed statistics
No changes to the training path or normalization math
No general video, timestamp, or dataset-quality validation
No requirement to run the model or a robot
Implementation and tests

I would be happy to implement this after confirming the preferred interface. The implementation could reuse the existing configuration, data-loader, and NormStats utilities while keeping the comparison logic in small pure functions.

Proposed tests:

Matching statistics produce a passing report
Dimension mismatches are reported clearly
NaN and Inf values are counted
Constant dimensions are identified
Out-of-range fractions are correct for a synthetic batch
The command remains read-only

Would maintainers prefer this as a separate check_norm_stats.py script or as a --compare-with-configured-stats mode in compute_norm_stats.py?

**Comment by sattyamjjain** [comment]
Supporting this from the evaluation side, where the same silent mismatch has a different and nastier symptom.

If normalization stats are wrong but syntactically valid, a policy still runs and still produces plausible-looking rollouts. What changes is the failure distribution, not the failure rate you notice. Anyone comparing two checkpoints, or comparing a perturbed run against a clean one, is then measuring a normalization difference and reporting it as a behavioural one.

A read-only report would fix that, and I would add one field to it: emit the resolved stats into the run artifact, not just to stdout. Then a result carries the normalization it was produced under, and a reviewer can check it months later without rerunning anything.

That single field is the difference between a number someone can audit and a number they have to trust.

**Comment by TheRealBaoBao** [comment]
Thanks—this is a great point. A normalization mismatch can become a hidden confounder when comparing policies, even when both evaluations run successfully.

I’ve already opened a pull request implementing the initial read-only compatibility report: https://github.com/Physical-Intelligence/openpi/pull/1026. I’d be happy to extend it with an optional machine-readable artifact containing the resolved normalization statistics and report metadata, while keeping the command fully read-only.

I’ll add that once the maintainers confirm the preferred artifact format and scope.

**Comment by sattyamjjain** [comment]
Correcting my own comment first: I said "silent", and openpi is better than that in the missing case. `transform_dataset` raises when `data_config.norm_stats is None`, and `normalize.load` raises `FileNotFoundError` on the serving path. Absent stats already fail loudly.

The gap is wrong-but-present stats. A valid `norm_stats.json` for the wrong robot loads cleanly, raises nothing, and trains. That is the case this PR catches, and I think it is worth scoping it that way in the PR description.

On the format question, I think the repo already answers it. `openpi.shared.normalize.serialize_json()` produces exactly this JSON today; it is what `normalize.save()` writes to `norm_stats.json`. No new format needed.

What cannot be reconstructed later is provenance, not the numbers. `create_base_config` resolves stats from `assets.assets_dir or assets_dirs` joined with `asset_id`, and that can be a `gs://` URI through `maybe_download`. `pi05_droid_finetune` pulls from `gs://openpi-assets/checkpoints/pi05_droid/assets` with `asset_id="droid"`. After the run, the local cache path does not tell you which remote asset produced it.

So, two fields rather than a schema: the resolved `assets_dir` and `asset_id`, plus a hash of the serialized stats. The hash is what lets someone compare two runs without diffing arrays.
